File: php/11-container-with-most-water.php
class Solution {

    /**
    * @param Integer[] $height
    * @return Integer
    */
    function maxArea($height) {
        $left = 0;
        $right = count($height)-1;
        $max = 0;
        while ($left < $right) {
            $candidate = $this->calculateArea($height, $left, $right);
            $max = max($max, $candidate);
            if ($height[$left] < $height[$right]) {
                $left++;
            } else {
                $right--;
            }
        }
        return $max;
    }

    function calculateArea($height, $left, $right) {
        return ($right - $left) * min($height[$left], $height[$right]);
    }
}

File: golang/11-container-with-most-water.go
package golang

func maxArea(height []int) int {
	calculateArea := func(left int, right int) int {
		x := right - left
		if height[left] > height[right] {
			return x * height[right]
		}
		return x * height[left]
	}

	left, right, max := 0, len(height)-1, 0
	for left < right {
		candidate := calculateArea(left, right)
		if candidate > max {
			max = candidate
		}
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}
	return max
}
